src/suites/cts/validation/commandBuffer.spec.ts

```diff
@@ -0,0 +1,281 @@
+export const description = `
+commandBuffer validation tests.
+`;
+
+import { TestGroup } from '../../../framework/index.js';
+
+import { ValidationTest } from './validation_test.js';
+
+export class F extends ValidationTest {
+  beginPass(
+    type: string,
+    commandEncoder: GPUCommandEncoder
+  ): GPURenderPassEncoder | GPUComputePassEncoder {
+    if (type === 'render') {
+      return this.beginRenderPass(commandEncoder);
+    } else if (type === 'compute') {
+      return this.beginComputePass(commandEncoder);
+    } else {
+      throw new Error('Unexpected pass encoder type');
+    }
+  }
+
+  beginRenderPass(commandEncoder: GPUCommandEncoder): GPURenderPassEncoder {
+    const attachmentTexture = this.device.createTexture({
+      format: 'rgba8unorm',
+      size: { width: 16, height: 16, depth: 1 },
+      usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+    });
+
+    return commandEncoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: attachmentTexture.createView(),
+          loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+        },
+      ],
+    });
+  }
+
+  beginComputePass(commandEncoder: GPUCommandEncoder): GPUComputePassEncoder {
+    return commandEncoder.beginComputePass();
+  }
+}
+
+export const g = new TestGroup(F);
+
+g.test('basic', async t => {
+  const commandEncoder = t.device.createCommandEncoder();
+  commandEncoder.finish();
+});
+
+g.test('command buffer cannot be ended mid pass', async t => {
+  const { type } = t.params;
+
+  {
+    // Control case
+    const commandEncoder = t.device.createCommandEncoder();
+    const pass = t.beginPass(type, commandEncoder);
+    pass.endPass();
+    commandEncoder.finish();
+  }
+  {
+    // Command buffer ended mid-pass.
+    await t.expectValidationError(() => {
+      const commandEncoder = t.device.createCommandEncoder();
+      t.beginPass(type, commandEncoder);
+      commandEncoder.finish();
+    });
+  }
+  {
+    // Trying to use encoder pass after finish() should fail too
+    await t.expectValidationError(async () => {
+      const commandEncoder = t.device.createCommandEncoder();
+      const pass = t.beginPass(type, commandEncoder);
+      commandEncoder.finish();
+
+      await t.expectValidationError(() => {
+        pass.endPass();
+      });
+    });
+  }
```

Comment:
I think it would be better to split these 3 into separate tests. The only annoying bit of this is they're separately parameterized, but fortunately in this case the parameters are very short. (and can be slightly shorter with `poptions('type', ['render', 'compute'])`). You can also store that params list off into a variable and reuse it, if you want.

Reply:
Done.